bootstrap.js

```diff
@@ -0,0 +1,137 @@
+  }
+
+  async shouldDownload (dest, hashUnpacked, hashAlgorithm) {
+    let buf
+    try {
+      buf = await fsReadFile(dest)
+    } catch (err) {
+      return true
+    }
+    if (hashUnpacked) {
+      this.emit('hash-check', { status: 'started', check: 'unpacked' })
+      this.#hashUnpackedActual = await this.getHash(buf, hashAlgorithm)
+      const hashMatch = this.#hashUnpackedActual === hashUnpacked
+      this.emit('hash-check', { status: 'finished', check: 'unpacked', result: hashMatch })
+      return !hashMatch
+    }
+  }
+
+  async download (src, hashDownloaded, hashAlgorithm) {
+    const response = await fetch(src, {
+      mode: 'cors'
+    })
+    if (!response.ok) {
+      this.cleanup()
+      throw new Error(`Bootstrap request faild: ${response.status} ${response.statusText}`)
+    }
+    const reader = response.body.getReader()
+    const contentLength = +response.headers.get('Content-Length')
+    let receivedLength = 0
+    let chunks = []
+
+    const read = async () => {
+      const {done, value} = await reader.read();
+      if (done) return
+      chunks.push(value);
+      receivedLength += value.length;
+      this.emit('download-progress', (receivedLength / contentLength * 100) | 0)
+      await read()
+    }
+    await read ()
+
+    const uint8data = new Uint8Array(receivedLength);
+    let position = 0;
+    for(let chunk of chunks) {
+      uint8data.set(chunk, position);
+      position += chunk.length;
+    }
+
+    if (hashDownloaded) {
+      this.emit('hash-check', { status: 'started', check: 'downloaded' })
+      const hashDownloadedActual = await this.getHash(uint8data, hashAlgorithm)
+      const hashMatch = hashDownloadedActual === hashDownloaded
+      this.emit('hash-check', { status: 'finished', check: 'downloaded', result: hashMatch })
+      if (!hashMatch) {
+        this.cleanup()
+        throw new Error(`Hash mismatch (downloaded): ${hashDownloadedActual} !== ${hashDownloaded}`)  
+      }
+    }
+
+    return uint8data
+  }
+
+  async unpack (buf, unpack, hashUnpacked, hashAlgorithm) {
+    if (typeof unpack !== 'function') {
+      return buf
+    }
+    this.emit('unpack', { status: 'started' })
+    let unpacked = unpack(buf)
+    if (isAsyncFunction(unpack) || isPromiseLike(unpacked)) {
+      unpacked = await unpacked
+    }
+    this.emit('unpack', { status: 'finished' })
+    if (hashUnpacked) {
+      this.emit('hash-check', { status: 'started', check: 'unpacked' })
+      this.#hashUnpackedActual ??= await this.getHash(unpacked, hashAlgorithm)
+      const hashMatch = this.#hashUnpackedActual === hashUnpacked
+      this.emit('hash-check', { status: 'finished', check: 'unpacked', result: hashMatch })
+      if (!hashMatch) {
+        this.cleanup()
+        throw new Error(`Hash mismatch (unpacked): ${this.#hashUnpackedActual} !== ${hashUnpacked}`)
+      }
+    }
+    return unpacked
+  }
+
+  cleanup () {
+    this.removeAllListeners()
+  }
+}
+
+// TODO: move to WebWorker?
```

Comment:
probably but we'll need `window._ipc` to tbe there

Reply:
hmm. it could cause a performance downgrade